Projects/Server/Network/TcpServer.cs

```diff
@@ -123,50 +140,131 @@ public static Socket CreateListener(IPEndPoint ipep)
         return null;
     }
 
+    private static async void ValidateSockets()
+    {
+        List<Task<Socket>> receiveTasks = new List<Task<Socket>>(1024);
+        while (_startValidating.WaitOne())
+        {
+            if (!World.Running || World.Loading)
+            {
+                continue;
+            }
+
+            // TODO: Max count
+            while (_validatingSocket.TryDequeue(out var socket))
+            {
+                receiveTasks.Add(ValidateSocketAsync(socket));
+            }
+
+            if (receiveTasks.Count == 0)
+            {
+                continue;
+            }
+
+            var sockets = await Task.WhenAll(receiveTasks);
+
+            for (int i = 0; i < sockets.Length; i++)
+            {
+                var socket = sockets[i];
+                if (socket != null)
+                {
+                    _validatedSocket.Enqueue(socket);
+                }
+            }
+
+            receiveTasks.Clear();
+
+            if (Core.Closing)
+            {
+                return;
+            }
+        }
+    }
+
+    private static async Task<Socket> ValidateSocketAsync(Socket socket)
+    {
+        var firstBytes = new byte[5];
+        try
+        {
+            await socket.ReceiveAsync(firstBytes, SocketFlags.Peek).WaitAsync(TimeSpan.FromMilliseconds(20));
```

Comment:
Timeout so short won't work

Reply:
After some testing, 30 second timeout (30000) seems better. You can work your way down from there.

Reply:
I would want the server to wait the least amount of time between accept and first packet to minimize threat surface from DDoS.

Reply:
You can close bad connections on the network/firewall side by changing the TCP timeouts there to a 1 to 2 second threshold. The thing with TCP, its error checking, so you want the network/firewalls layer to cut off bad connections so MUO isn't doing it.

Reply:
I agree that this should definitely be configured outside of the game server. I believe the time out is set by the socket options as well in the code. And by default it never times out. I'm not sure if this overrides the operating system though?